staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+// ready is a condition variable that represent readiness and
+// allow callers to wait on its status, returning an error with the reason.
+// It allows to use context cancellation and to be stopped,
+// on both cases it stops waiting raising the corresponding error.
+type ready struct {
+	mu      sync.Mutex
+	readyCh chan struct{}
+	stopCh  chan struct{}
+}
+
+// newReady creates a ready condition variable that is not stopped and not ready
+func newReady() *ready {
+	return &ready{
+		readyCh: make(chan struct{}), // is ready if it does not exist
+		stopCh:  make(chan struct{}), // is stopped if it does not exist
+	}
+}
+
+// wait blocks until:
+// - it is ready and returns nil
+// - it is stopped and returns an error
+// - context is cancelled and returns error
+func (r *ready) wait(ctx context.Context) error {
+	select {
+	// if it is ready but context is cancelled
+	// context cancellation fails fast.
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		// readyCh and stopCh must not be closed at the same time
+		select {
+		case <-r.readyCh:
```

Comment:
Hmm - I started wondering if this function is actually thread-safe.
Because unset (in line 106) is actually modifies that.

It seems that your tests do check concurrent calls for set(), but not for unset() ...

Reply:
This channel is the synchronization signal , on `check` we avoid manipulating the state meanwhile we are checking the condition but on `wait` ... I can't see other way to implement this, you can't hold a lock meanwhile you are waiting, sync.Cond AFAIK operates the same, once it Wait it releases the lock and has to check the status again after it receives the .Signal and wakes up, since we use a channel to coordinate , it dictates the status inside the select clause

Reply:
sync.Cond (the original approach) didn't have this problem as we were checking the env var from scratch (it internally releases the lock as a last thing before going to idle mode).

The only way I can see is to switch back to the original sync.Cond approach but:
1) adding support for stopCh is easy - we can just have the stop() method that will adjust the value of added stopped field and signal all waiters
2) but I don't see an easy way for honoring context then without starting an additional goroutine for every waiter.

OTOH - we would have to start this goroutine only if the fast path of checking ready is false, so that would generally not fire, so maybe that's actually ok and good enough?

Reply:
SGTM, let's go simple and solve the stop problem now ... the context was a nice to have but is not real a problem now